File: src/ModelLib/Solver2.java
package ModelLib;

import java.util.Vector;

import static ModelLib.Utils.*;

public class Solver2 extends Solver
{
    public Solver2()
    {}

    public Solver2(Model m, Graph g, double e)
    {
        super(m, g, e);
    }

    @Override
    public void init(Vector<Integer> u, int l)
    {
        level = l;
        used = u;
        num = 2;
        reps.ensureCapacity(num);
        prob.ensureCapacity(num);
        sz = used.size();
        for (int i = 0; i < num; ++i)
            prob.get(i).ensureCapacity(sz);
        localMaxD = maxDistance();
        if (level == 0)
            Utils.globalAvgD = avgDistance();
        localAvgD = avgDistance();
        angRange = angDifference();
    }

    @Override
    public void expand()
    {
        double maxD = maxPatchDistance();
        System.out.println("minPatchD = " + maxD / globalMaxD);
        if (level > 2 || maxD / globalMaxD < Utils.TH_REP_DIST_RATIO) return;
        for (int i = 0; i < num; ++i)
        {
            Solver s = new Solver2(pModel, pGraph, eps);
            Vector<Integer> uNew = new Vector<>();
            for (int j = 0; j < sz; ++j)
                if ((pModel.faces.get(used.get(j)).label)%num == i)
                    uNew.add(used.get(j));
            s.init(uNew, level + 1);
            sons.add(s);
        }
        for (int i = 0; i < num; ++i)
        {
            System.out.println("maxD = " + sons.get(i).localMaxD / globalMaxD
                            + " angD = " + sons.get(i).angRange);

            if (sons.get(i).localMaxD / globalMaxD > TH_DIST_RATIO &&
                sons.get(i).angRange > TH_ANGLE_DIFF)
            sons.get(i).process();
        }
    }
}

File: src/Solver2.cpp
//
// Created by timofey on 28.09.2021.
//
#include <iostream>

#include "../include/Solver2.h"
#include "../include/Util.h"

using namespace std;
using namespace util;

void Solver2::init(vector<int>& u, int l)
{
    level = l;
    used = u;
    num = 2;
    reps.resize(num);
    prob.resize(num);
    sz = used.size();
    for (int i = 0; i < num; ++i)
        prob[i].resize(sz);
    localMaxD = maxDistance();
    if (level == 0)
        globalAvgD = avgDistance();
    localAvgD = avgDistance();
    angRange = angDifference();
}

void Solver2::expand()
{
    double maxD = maxPatchDistance();
    cout << "minPatchD = " << maxD / globalMaxD << endl;
    if (level > 2 || maxD / globalMaxD < TH_REP_DIST_RATIO) return;
    for (int i = 0; i < num; ++i)
    {
        Solver* s = new Solver2(pModel, pGraph, eps);
        vector<int> uNew;
        for (int j = 0; j < sz; ++j)
            if ((pModel->faces[used[j]].label)%num == i)
                uNew.push_back(used[j]);
        s->init(uNew, level + 1);
        sons.push_back(s);
    }
    for (int i = 0; i < num; ++i)
    {
        cout << "maxD = " << sons[i]->localMaxD / globalMaxD
             << " angD = " << sons[i]->angRange << endl;
        if (sons[i]->localMaxD / globalMaxD > TH_DIST_RATIO &&
            sons[i]->angRange > TH_ANGLE_DIFF)
            sons[i]->process();
    }
}
